### DISCUSSION TITLE
Document librespot abilities - how to do playback ?

### DISCUSSION THREAD

**Discussion by GChalony** [librespot-org/librespot#889] [General]
Hi, I'd like to know if there is anywhere I can find some documentation about what librespot does exactly. I think it's a bit misleading that the sample program is a Spotify Connect receiver because it suggests that it's what `librespot` is all about. If I understand correctly it does much more, but I couldn't find an API reference or a tutorial to explain what can be done with it.

I'd personally like to play Spotify albums based on scanned NFC tags, so I'm just looking for a way to play the songs based on a URI.

I'm also wondering how `librespot` works, because I've seen that the `Spotify Web API` doesn't allow playback on local devices. `libspotify` seemed to be able to do it but has been deprecated for a long time.

Sorry if I'm a bit lost, that a lot of new stuff for me and it can be a bit overwhelming. Cheers!

**Comment by ashthespy** [comment]
Indeed, the primary usecase is to provide a connect endpoint. 

But, you can also pass in a uri and have it play, by using `librespot` as a library. Have a look at the [examples](https://github.com/librespot-org/librespot/tree/dev/examples) folder on how this might be done! :-)

**Comment by GChalony** [comment]
I see, thanks :) I already had a look at the example but to be honest I don't know how to run them. What would you recommend be the best way to learn how to implement my program ? Looking directly at the source code ?

**Reply by kingosticks** [reply]
Did you try something like https://stackoverflow.com/questions/54469463/how-do-i-run-a-projects-example-using-cargo ?

**Reply by GChalony** [reply]
I'll give it a shot, seems to fail with 
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: MercuryError', examples/playlist_tracks.rs:31:58`
```
 for now

**Reply by kingosticks** [reply]
What playlists did you try?

**Reply by GChalony** [reply]
My bad I was actually trying a song. Works fine, it's great! So just to confirm, is there anywhere I can learn how to use the library (maybe in `librespot-java`?) or is it better to just jump into the code ?

**Reply by ashthespy** [reply]
What are you trying to achieve? Is your goal to learn Rust? Or if you just want a player, look at projects that wrap librespot, such as spotidyd.

PS: librespot-java is on different, newer API

**Reply by GChalony** [reply]
I really only need a player, though I'm not entirely sure what it is exactly. If I understand correctly, there needs to be a "device" (a spotify client?) running on some computer/whatever, and then the `Spotify Web API` can be used to control what is being played on that device. `spotifyd` (powered by `librespot`) is such a "device" so I can use any tool to control it with the Web API, like `spotify-tui`.  
I am close ? :sweat_smile:
